documentation/development-docs/patterns-for-standard-cmdlets.md

```diff
@@ -0,0 +1,996 @@
+
+<details><summary>Click to expand example</summary>
+<p>
+
+```cs
+namespace Microsoft.Azure.Commands.Foo
+{
+    [Cmdlet(VerbsCommon.Get, "AzureRmFoo", DefaultParameterSetName = "ListParameterSet"), OutputType(typeof(PSFoo))]
+    public class GetFooCommand : FooBaseCmdlet
+    {
+        [Parameter(Mandatory = false, ParameterSetName = "ListParameterSet")]
+        [Parameter(Mandatory = true, ParameterSetName = "GetByFooNameParameterSet")]
+        [ResourceGroupCompleter]
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+
+        [Parameter(Mandatory = true, ParameterSetName = "GetByFooNameParameterSet")]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
+
+        [Parameter(Mandatory = true, ValueFromPipelineByPropertyName = true, ParameterSetName = "GetByResourceIdParameterSet" )]
+        [ValidateNotNullOrEmpty]
+        public string ResourceId { get; set; }
+
+        public override void ExecuteCmdlet()
+        {
+            if (this.IsParameterBound(c => c.ResourceId))
+            {
+                var resourceIdentifier = new ResourceIdentifier(this.ResourceId);
+                this.ResourceGroupName = resourceIdentifier.ResourceGroupName;
+                this.Name = resourceIdentifier.ResourceName;
+            }
+
+            if (!string.IsNullOrEmpty(this.Name))
+            {
+                var result = new PSFoo(this.MySDKClient.Foo.Get(this.ResourceGroupName, this.Name));
+                WriteObject(result);
+            }
+            else if (!string.IsNullOrEmpty(this.ResourceGroupName))
+            {
+                var result = this.MySDKClient.Foo.ListByResourceGroup(this.ResourceGroupName).Select(f => new PSFoo(f));
+                WriteObject(result, true);
+            }
+            else
+            {
+                var result = this.MySDKClient.Foo.List().Select(f => new PSFoo(f));
+                WriteObject(result, true);
+            }
+        }
+    }
+}
+```
+
+</p>
+</details>
+
+### `New-*` cmdlet
+
+All top-level resources should have a `New-*` cmdlet that allows users to create a resource with given properties. This cmdlet should implement `SupportsShouldProcess` to allow users to provide the `-WhatIf` parameter and see what the result of executing the cmdlet is without the resource actually being created.
+
+#### Parameter sets
+
+To enable the above scenario, only one parameter set is needed:
+
+```
+New-AzureRmFoo -ResourceGroupName <String> -Name <String> [-Property1 <Type1>] [-Property2 <Type2>] ... [-WhatIf] [-Confirm]
+```
+
+This parameter set has required `-ResourceGroupName` and `-Name` parameters to satisfy the identity properties of the resource, as well as a few optional `-PropertyX` parameters that allows the user to set values for properties. The parameter set also has optional `-WhatIf` and `-Confirm` parameters that are automatically included from the implementation of `SupportsShouldProcess`.
```

Comment:
Note that all properties required by the API should be required here.  This may involve using additional parameter sets for different kinds of resources (for example, linux and windows VMs will have different required properties, and so should use additional parameter sets)

Reply:
Added a comment about this.